src/NzbDrone.Core/Notifications/Webhook/WebhookService.cs

```diff
@@ -0,0 +1,113 @@
+﻿using FluentValidation.Results;
+using NLog;
+using NzbDrone.Common.Processes;
+using NzbDrone.Core.MediaFiles;
+using NzbDrone.Core.Tv;
+using NzbDrone.Core.Validation;
+using NzbDrone.Core.Rest;
+using RestSharp;
+using NzbDrone.Core.Qualities;
+using NzbDrone.Core.Parser.Model;
+
+namespace NzbDrone.Core.Notifications.Webhook
+{
+    public interface IWebhookService
+    {
+        void OnDownload(Series series, EpisodeFile episodeFile, WebhookSettings settings);
+        void OnRename(Series series, WebhookSettings settings);
+        void OnGrab(Series series, RemoteEpisode episode, QualityModel quality, WebhookSettings settings);
+        ValidationFailure Test(WebhookSettings settings);
+    }
+
+    public class WebhookService : IWebhookService
+    {
+        private readonly IProcessProvider _processProvider;
+        private readonly Logger _logger;
+
+        public WebhookService(IProcessProvider processProvider, Logger logger)
+        {
+            _processProvider = processProvider;
+            _logger = logger;
+        }
+
+        public void OnDownload(Series series, EpisodeFile episodeFile, WebhookSettings settings)
+        {
+            WebhookPayload payload = new WebhookPayload
+            {
+                EventType = "Download",
+                Series = new WebhookSeries(series),
+                Episodes = episodeFile.Episodes.Value.ConvertAll(x => new WebhookEpisode(x) {
+                    Quality = episodeFile.Quality.Quality.Name,
+                    QualityVersion = episodeFile.Quality.Revision.Version,
+                    ReleaseGroup = episodeFile.ReleaseGroup,
+                    SceneName = episodeFile.SceneName
+                })
+            };
+
+            NotifyWebhook(payload, settings);
+        }
+
+        public void OnRename(Series series, WebhookSettings settings)
+        {
+            WebhookPayload payload = new WebhookPayload
+            {
+                EventType = "Rename",
+                Series = new WebhookSeries(series)
+            };
+
+            NotifyWebhook(payload, settings);
+        }
+
+        public void OnGrab(Series series, RemoteEpisode episode, QualityModel quality, WebhookSettings settings)
+        {
+            WebhookPayload payload = new WebhookPayload
+            {
+                EventType = "Grab",
+                Series = new WebhookSeries(series),
+                Episodes = episode.Episodes.ConvertAll(x => new WebhookEpisode(x)
+                {
+                    Quality = quality.Quality.Name,
+                    QualityVersion = quality.Revision.Version,
+                    ReleaseGroup = episode.ParsedEpisodeInfo.ReleaseGroup
+                })
+            };
+            NotifyWebhook(payload, settings);
+        }
+
+        public void NotifyWebhook(object body, WebhookSettings settings)
```

Comment:
I think that would make more sense, the other option which I think would be ok is to add a default constructor to WebhookPayload` and all it's subclasses, that way you can create a custom one rather than always having to create them using `Series` etc.